### PR TITLE
Fixes #15402 - Moved puppet to separate api controller

### PR DESCRIPTION
**Pull request by ShimShtein**

### CODE REVIEW

**Comment by ShimShtein** [comment]
The tests hit `Execution expired` error, retrying.
[test]

**Comment by dLobatog** [comment]
@ShimShtein Thanks, most of it is good :smile:  Just two comments that I think would improve this:
1. Can you add a test for `permissions_check`? Just to check we call it properly on the puppetrun action.
2. There's an additional route created 'api/v2/puppet_hosts/puppetrun', can you make this patch not add any additional routes? I don't think it's used for anything.

**Comment by ShimShtein** [comment]
Added test for `permissions_check`.

> There's an additional route created 'api/v2/puppet_hosts/puppetrun', can you make this patch not add any additional routes? I don't think it's used for anything.

@dLobatog , are you sure? I don't see it when running `rake routes`:

```
$ rake routes | grep puppetrun
                                                   puppetrun_host GET    /hosts/:id/puppetrun(.:format)                                                                                                                         hosts#puppetrun {:id=>/[^\/]+/}
                                         multiple_puppetrun_hosts GET    /hosts/multiple_puppetrun(.:format)                                                                                                                    hosts#multiple_puppetrun
                                  update_multiple_puppetrun_hosts POST   /hosts/update_multiple_puppetrun(.:format)                                                                                                             hosts#update_multiple_puppetrun
                                               puppetrun_api_host PUT    /api(/:apiv)/hosts/:id/puppetrun(.:format)                                                                                                             api/v2/puppet_hosts#puppetrun {:format=>"json", :apiv=>/v1|v2/, :id=>/[^\/]+/}
```

in the last line I can see the new method, but the path is set to `/api/v2/hosts/:id/puppetrun`.

**Comment by ShimShtein** [comment]
Test results are inconsistent, retesting.
[test]

**Comment by ShimShtein** [comment]
Rebased.

**Comment by ShimShtein** [comment]
[test]

File: test/lib/foreman/access_permissions_test.rb [hunk 1]
```diff
@@ -44,6 +44,8 @@ class AccessPermissionsTest < ActiveSupport::TestCase
 
   MAY_SKIP_AUTHORIZED = [ "about/index" ]
 
+  SPECIAL_PATH = ['api/v2/puppet_hosts/puppetrun']
```
Comment (orrabin) [inline test/lib/foreman/access_permissions_test.rb]: maybe change to PUPPET_RUN_PATH or HOSTS_PUPPET_RUN_PATH?

Comment (ShimShtein) [reply test/lib/foreman/access_permissions_test.rb]: It's a global constant that holds all actions that have special paths - these actions cannot be tested by this test. [This line](https://github.com/ShimShtein/foreman/blob/73fe3dc47c3faee6d44ced1f059bf9cd5dce176d/test/lib/foreman/access_permissions_test.rb#L51) generates paths from controller names, and in this case it's not going to work.

**Comment by ShimShtein** [comment]
Rebased

**Comment by ShimShtein** [comment]
Rebased
@dLobatog, Can you please re-review this?

**Comment by mmoll** [comment]
this needs a rebase :)

**Comment by ShimShtein** [comment]
Rebased

**Comment by iNecas** [comment]
Tests are green. Looks good to me. @dLobatog anything else to add before we merge?

**Review by dLobatog — CHANGES_REQUESTED** [review]
@ShimShtein @iNecas This looks (and looked) basically ready to merge to me, it slipped off my radar. Are you able to rebase so we can see tests pass & we move forward?

Sorry for the long wait to get back to you

**Comment by ShimShtein** [comment]
Rebased, tests are green. @dLobatog want to take last round?

**Review by dLobatog — APPROVED** [review]

**Comment by dLobatog** [comment]
Thanks @ShimShtein !

**Pull request merged**